### ISSUE TITLE
Docs generator drops mandatory parentheses around Proc notations

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#9964]
Parentheses around `->` literals are mandatory in at least 2 places:

* Nested `Proc`s, i.e. when any argument or return type of a `Proc` is itself a `Proc`:
  ```crystal
  def foo(func : Int32 -> (String -> Char))
    foo(->(x : Int32) { ->(y : String) { 'a' } })
  end

  def bar(func : (Int32 -> String) -> Char)
    bar(->(x : Int32 -> String) { 'b' })
  end
  ```
  Both methods currently generate `Int32 -> String -> Char`, which is invalid Crystal code.
* Nilable `Proc`s with any argument or return type (`->?` alone is okay):
  ```crystal
  def foo(func : -> Int32?)
    foo(->{ 1 })
    foo(->{ })
  end

  def bar(func : (-> Int32)?)
    bar(->{ 1 })
    bar(nil)
  end
  ```
  `bar` generates `-> Int32?`, the same as `foo`. One example in the standard library is the following [constructor of `Hash`](https://crystal-lang.org/api/0.35.1/Hash.html#new(block:Hash(K,V),K-%3EV?=nil,*,initial_capacity=nil)-class-method):

  ```crystal
  def self.new(block : Hash(K, V), K -> V? = nil, *, initial_capacity = nil) 
  ```

  whose actual signature is:

  ```crystal
  def initialize(block : (Hash(K, V), K -> V)? = nil, *, initial_capacity = nil)
  ```

**Comment by bcardiff** [comment]
Note: this is likely a bug in `ToSVisitor`

**Comment by HertzDevil** [comment]
The docs generator doesn't depend on `ToSVisitor`, as the typenames are generated here:

https://github.com/crystal-lang/crystal/blob/1a97e71926143cfed8bcd7acde1c0655213d5059/src/compiler/crystal/tools/doc/type.cr#L505-L515

https://github.com/crystal-lang/crystal/blob/1a97e71926143cfed8bcd7acde1c0655213d5059/src/compiler/crystal/tools/doc/type.cr#L588-L595

The "always put parentheses around `->` literals" fix is simply putting `io << '('` and `io << ')'` around these methods.

**Comment by asterite** [comment]
I don't think it will look nice if we always put parentheses around these types. It should probably be done only when necessary.
